packages/datetime/src/datePickerCaption.tsx

```diff
@@ -123,25 +104,29 @@ export class DatePickerCaption extends React.PureComponent<IDatePickerCaptionPro
         this.positionArrows();
     }
 
-    private containerRefHandler = (r: HTMLElement) => (this.containerElement = r);
-
     private positionArrows() {
         // measure width of text as rendered inside our container element.
-        const monthWidth = Utils.measureTextWidth(
+        const monthWidth = measureTextWidth(
             this.displayedMonthText,
             Classes.DATEPICKER_CAPTION_MEASURE,
             this.containerElement,
         );
-        this.setState({ monthWidth });
+        if (monthWidth > this.containerElement.firstElementChild.clientWidth - Icon.SIZE_STANDARD) {
```

Comment:
~Can you post some screenshots here to demonstrate each case by example?~ Never mind.